azure/services/scalesets/spec.go

```diff
@@ -0,0 +1,522 @@
+	DiagnosticsProfile           *infrav1.Diagnostics
+	FailureDomains               []string
+	VMExtensions                 []infrav1.VMExtension
+	NetworkInterfaces            []infrav1.NetworkInterface
+	IPv6Enabled                  bool
+	OrchestrationMode            infrav1.OrchestrationModeType
+	Location                     string
+	SubscriptionID               string
+	SKU                          resourceskus.SKU
+	VMSSExtensionSpecs           []azure.ResourceSpecGetter
+	VMImage                      *infrav1.Image
+	BootstrapData                string
+	VMSSInstances                []compute.VirtualMachineScaleSetVM
+	MaxSurge                     int
+	ClusterName                  string
+	ShouldPatchCustomData        bool
+	HasReplicasExternallyManaged bool
+	AdditionalTags               infrav1.Tags
+}
+
+// ResourceName returns the name of the subnet.
+func (s *ScaleSetSpec) ResourceName() string {
+	return s.Name
+}
+
+// ResourceGroupName returns the name of the resource group of the VNet that owns this subnet.
+func (s *ScaleSetSpec) ResourceGroupName() string {
+	return s.ResourceGroup
+}
+
+// OwnerResourceName returns the name of the VNet that owns this subnet.
+func (s *ScaleSetSpec) OwnerResourceName() string {
+	return ""
+}
+
+func (s *ScaleSetSpec) existingParameters(ctx context.Context, existing interface{}) (parameters interface{}, err error) {
+	existingVMSS, ok := existing.(compute.VirtualMachineScaleSet)
+	if !ok {
+		return nil, errors.Errorf("%T is not a compute.VirtualMachineScaleSet", existing)
+	}
+
+	existingInfraVMSS := converters.SDKToVMSS(existingVMSS, s.VMSSInstances)
+
+	params, err := s.Parameters(ctx, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to generate scale set update parameters for %s", s.Name)
+	}
+
+	vmss, ok := params.(compute.VirtualMachineScaleSet)
+	if !ok {
+		return nil, errors.Errorf("%T is not a compute.VirtualMachineScaleSet", existing)
+	}
+
+	vmss.VirtualMachineProfile.NetworkProfile = nil
+	vmss.ID = existingVMSS.ID
+
+	hasModelChanges := hasModelModifyingDifferences(existingInfraVMSS, vmss)
+	isFlex := s.OrchestrationMode == infrav1.FlexibleOrchestrationMode
+	updated := true
+	if !isFlex {
+		updated = existingInfraVMSS.HasEnoughLatestModelOrNotMixedModel()
+	}
+	if s.MaxSurge > 0 && (hasModelChanges || !updated) && !s.HasReplicasExternallyManaged {
+		// surge capacity with the intention of lowering during instance reconciliation
+		surge := s.Capacity + int64(s.MaxSurge)
+		vmss.Sku.Capacity = pointer.Int64(surge)
+	}
+
+	// If there are no model changes and no increase in the replica count, do not update the VMSS.
+	// Decreases in replica count is handled by deleting AzureMachinePoolMachine instances in the MachinePoolScope
+	if *vmss.Sku.Capacity <= existingInfraVMSS.Capacity && !hasModelChanges && !s.ShouldPatchCustomData {
+		// up to date, nothing to do
+		return nil, nil
+	}
+
+	return vmss, nil
+}
+
+// Parameters returns the parameters for the subnet.
```

Comment:
Same comment as above

Reply:
this is still valid, should be

```suggestion
// Parameters returns the parameters for the Scale Set.
```

Reply:
this is resolved